crates/noirc_evaluator/src/ssa/node.rs

```diff
@@ -0,0 +1,948 @@
+    fn get_type(&self) -> ObjectType {
+        match self {
+            NodeObj::Obj(o) => o.get_type(),
+            NodeObj::Instr(i) => i.res_type,
+            NodeObj::Const(o) => o.value_type,
+        }
+    }
+
+    fn print(&self) -> String {
+        match self {
+            NodeObj::Obj(o) => o.print(),
+            NodeObj::Instr(i) => i.print_i(),
+            NodeObj::Const(c) => c.print(),
+        }
+    }
+
+    fn bits(&self) -> u32 {
+        match self {
+            NodeObj::Obj(o) => o.bits(),
+            NodeObj::Instr(i) => i.res_type.bits(),
+            NodeObj::Const(c) => c.bits(),
+        }
+    }
+
+    fn get_id(&self) -> arena::Index {
+        match self {
+            NodeObj::Obj(o) => o.get_id(),
+            NodeObj::Instr(i) => i.idx,
+            NodeObj::Const(c) => c.get_id(),
+        }
+    }
+}
+
+impl Node for Constant {
+    fn get_type(&self) -> ObjectType {
+        self.value_type
+    }
+
+    fn print(&self) -> String {
+        self.value.to_string()
+    }
+
+    fn bits(&self) -> u32 {
+        self.value.bits().try_into().unwrap()
+    }
+
+    fn get_id(&self) -> arena::Index {
+        self.id
+    }
+}
+
+#[derive(Debug)]
+pub enum NodeObj {
+    Obj(Variable),
+    Instr(Instruction),
+    Const(Constant),
+}
+
+#[derive(Debug)]
+pub struct Constant {
+    pub id: arena::Index,
+    pub value: BigUint,    //TODO use FieldElement instead
+    pub value_str: String, //TODO ConstStr subtype
+    pub value_type: ObjectType,
+}
+
+#[derive(Debug)]
+pub struct Variable {
+    pub id: arena::Index,
+    pub obj_type: ObjectType,
+    pub name: String,
+    //pub cur_value: arena::Index, //for generating the SSA form, current value of the object during parsing of the AST
+    pub root: Option<arena::Index>, //when generating SSA, assignment of an object creates a new one which is linked to the original one
+    pub def: Option<IdentId>,       //TODO redondant with root - should it be an option?
+    //TODO clarify where cur_value and root is stored, and also this:
+    //  pub max_bits: u32,                  //max possible bit size of the expression
+    //  pub max_value: Option<BigUInt>,     //maximum possible value of the expression, if less than max_bits
+    pub witness: Option<Witness>,
+    pub parent_block: arena::Index,
+}
+
+impl Variable {
+    pub fn get_root(&self) -> arena::Index {
+        match self.root {
+            Some(r) => r,
+            _ => self.id,
+        }
+    }
+}
+
+#[allow(non_camel_case_types)]
+#[derive(Copy, Clone, Debug)]
+pub enum ObjectType {
+    native_field,
+    // custom_field(BigUint), //TODO requires copy trait for BigUint
+    boolean,
+    unsigned(u32), //bit size
+    signed(u32),   //bit size
+    //custom(u32),   //user-defined struct, u32 refers to the id of the type in...?todo
+    //array(ObjectType),  TODO we should have primitive type and composite types
+    //TODO big_int
+    //TODO floats
+    none, //not an object
```

Comment:
Can we just use an Option instead of having this none variant?

Reply:
We could but a few instructions have no type (e.g jump) and if we use an option for instruction type, it will imply many checks that for most instructions are not needed.
I guess I could use any existing type for the instructions that do not have type since it won't be used but I think it is better to make it explicit like this.

Reply:
Okay that sounds fair, I don't have a strong opinion on this, so will leave it